Modules/_collectionsmodule.c

```diff
@@ -1050,39 +1071,52 @@ deque_index(dequeobject *deque, PyObject *const *args, Py_ssize_t nargs)
         start = stop;
     assert(0 <= start && start <= stop && stop <= Py_SIZE(deque));
 
-    for (i=0 ; i < start - BLOCKLEN ; i += BLOCKLEN) {
-        b = b->rightlink;
-    }
-    for ( ; i < start ; i++) {
-        index++;
-        if (index == BLOCKLEN) {
-            b = b->rightlink;
-            index = 0;
-        }
-    }
+    index = start;
+    b = deque_get_block(deque, &index);
 
-    n = stop - i;
+    n = stop - start ;
     while (--n >= 0) {
         CHECK_NOT_END(b);
         item = b->data[index];
-        cmp = PyObject_RichCompareBool(item, v, Py_EQ);
+        cmp = PyObject_RichCompareBool(item, value, Py_EQ);
         if (cmp > 0)
-            return PyLong_FromSsize_t(stop - n - 1);
+            return stop - n - 1;
         if (cmp < 0)
-            return NULL;
+            return -1;
         if (start_state != deque->state) {
             PyErr_SetString(PyExc_RuntimeError,
                             "deque mutated during iteration");
-            return NULL;
+            return -1;
         }
-        index++;
-        if (index == BLOCKLEN) {
+
+        if (++index == BLOCKLEN) {
             b = b->rightlink;
             index = 0;
         }
     }
-    PyErr_Format(PyExc_ValueError, "%R is not in deque", v);
-    return NULL;
+    PyErr_Format(PyExc_ValueError, "%R is not in deque", value);
+    return -1;
+}
+
+
+static PyObject *
+deque_index(dequeobject *deque, PyObject *args, Py_ssize_t nargs)
+{
+    Py_ssize_t i, start=0, stop=Py_SIZE(deque);
+    PyObject *value;
+
+    if (!_PyArg_ParseStack(args, nargs, "O|O&O&:index", &value,
+                           _PyEval_SliceIndexNotNone, &start,
+                           _PyEval_SliceIndexNotNone, &stop)) {
+        return NULL;
+    }
+
+    i = _deque_index(deque, value, start, stop);
+
+    if (i == -1 && PyErr_Occurred()) {
```

Comment:
Since `-1` always indicates an error is set, you can skip the ` && PyErr_Occurred()` and instead put an assertion inside the if-body.